spring-boot-project/spring-boot/src/main/java/org/springframework/boot/webservices/client/WebServiceTemplateBuilder.java

```diff
@@ -0,0 +1,1106 @@
+				this.webServiceMessageSenderCustomizers, this.marshaller,
+				this.unmarshaller, this.destinationProvider, this.transformerFactoryClass,
+				this.messageFactory, this.detectWebServiceMessageSender);
+	}
+
+	/**
+	 * Set {@link WebServiceTemplateCustomizer WebServiceTemplateCustomizers} that should
+	 * be applied to the {@link WebServiceTemplate}. Customizers are applied in the order
+	 * that they were added after builder configuration has been applied.
+	 *
+	 * <b>Note!</b> Override existing customizers
+	 * @param customizers the customizers to set
+	 * @return a new builder instance
+	 */
+	public WebServiceTemplateBuilder setCustomizers(
+			WebServiceTemplateCustomizer... customizers) {
+		Assert.notNull(customizers, "customizers must not be null");
+		return setCustomizers(Arrays.asList(customizers));
+	}
+
+	/**
+	 * Add additional {@link WebServiceTemplateCustomizer WebServiceTemplateCustomizers}
+	 * that should be applied to the {@link WebServiceTemplate}. Customizers are applied
+	 * in the order that they were added after builder configuration has been applied.
+	 * @param customizers the customizers to add
+	 * @return a new builder instance
+	 */
+	public WebServiceTemplateBuilder addCustomizers(
+			WebServiceTemplateCustomizer... customizers) {
+		Assert.notNull(customizers, "customizers must not be null");
+		return addCustomizers(Arrays.asList(customizers));
+	}
+
+	/**
+	 * Add additional {@link WebServiceTemplateCustomizer WebServiceTemplateCustomizers}
+	 * that should be applied to the {@link WebServiceTemplate}. Customizers are applied
+	 * in the order that they were added after builder configuration has been applied.
+	 * @param customizers the customizers to add
+	 * @return a new builder instance
+	 */
+
+	public WebServiceTemplateBuilder addCustomizers(
+			Collection<? extends WebServiceTemplateCustomizer> customizers) {
+		Assert.notNull(customizers, "customizers must not be null");
+		return new WebServiceTemplateBuilder(this.interceptors, this.internalCustomizers,
+				append(this.customizers, customizers),
+				this.webServiceMessageSenderSuppliers,
+				this.webServiceMessageSenderCustomizers, this.marshaller,
+				this.unmarshaller, this.destinationProvider, this.transformerFactoryClass,
+				this.messageFactory, this.detectWebServiceMessageSender);
+	}
+
+	/**
+	 * Sets the {@code Suppliers} of {@link WebServiceMessageSender} that should be called
+	 * each time when {@link #configure(WebServiceTemplate)} method is called.
+	 *
+	 * <b>Note!</b> Override existing WebServiceMessageSender {@code suppliers}
+	 * @param webServiceMessageSenderSuppliers Suppliers for the messageSenders
+	 * @return a new builder instance.
+	 * @see WebServiceTemplate#setMessageSenders(WebServiceMessageSender[])
+	 */
+
+	public WebServiceTemplateBuilder setWebServiceMessageSenders(
+			Collection<? extends Supplier<? extends WebServiceMessageSender>> webServiceMessageSenderSuppliers) {
```

Comment:
I've removed the Supplier in my polish commit. I didn't found a single use of it and would argue that it is something you can determine when you build the template

Reply:
You can't do this because of 'setReadTimeout' and 'setConnectionTimeout'. This test will not work.
```
HttpUrlConnectionMessageSender sender = new HttpUrlConnectionMessageSender();

		 WebServiceTemplate ws = this.builder
				 .setConnectionTimeout(5000)
				 .setReadTimeout(2000)
				 .messageSenders(sender)
				 .build();

		WebServiceTemplate ws1 = this.builder
				.setConnectionTimeout(4000)
				.setReadTimeout(3000)
				.build();

		assertThat(ReflectionTestUtils.getField(ws.getMessageSenders()[0], "connectionTimeout"))
				.isEqualTo(Duration.ofMillis(5000));
		assertThat(ReflectionTestUtils.getField(ws.getMessageSenders()[0], "readTimeout"))
				.isEqualTo(Duration.ofMillis(2000));
		//

		assertThat(ReflectionTestUtils.getField(ws1.getMessageSenders()[0], "connectionTimeout")).isEqualTo(Duration.ofMillis(4000));
		assertThat(ReflectionTestUtils.getField(ws1.getMessageSenders()[0], "readTimeout"))
				.isEqualTo(Duration.ofMillis(3000));

```

Reply:
IMO that's one more argument to remove that feature.

Reply:
Only for custom `WebServiceMessagaSenders`